### PR TITLE
Documentation for the timestamp module

### PR DESCRIPTION
**Pull request by gautamdhameja**
* Added code comments in the lib.rs of the timestamp module
* Added readme.md in the root of timestamp module crate. This file has the module docs as per the SRML documentation template.

### CODE REVIEW

**Comment by Kianenigma** [comment]
Whenever I wanted to learn about a new srml, lack of explanation on the test cases was quite a barrier. I would recommend putting some explanatory docs on the tests as well since they _should_ cover the common use cases of the module + common error/edge cases. Hence, I think they are quite important and should be documented + potentially linked as a further source of learning at the end of the .md file.

**Comment by ltfschoen** [comment]
Is there any opportunity to incorporate: https://github.com/paritytech/substrate/blob/master/README.adoc#72-contributing-to-documentation-for-substrate-packages

File: srml/timestamp/readme.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+### Types
+
+* `Moment` - Represents the current timestamp. Generally implemented as a `u64`.
+
+### Storage Items
+
+* `Now`: `Moment` - The current timestamp. A `u64` having the timestamps as **total seconds from the unix epoch**.
+* `BlockPeriod`: `Moment` - The minimum (and advised) period between blocks. Also represented as `Moment`.
+* `DidUpdate`: `bool` - Private flag to check if the timestamp was updated for the current block.
```
Comment (bkchr) [inline srml/timestamp/readme.md]: General note, as these are private and implementation details, what is the benefit for the user?

File: srml/timestamp/readme.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+### Public Inspection functions - Immutable (getters)
+
+#### get()
```
Comment (bkchr) [inline srml/timestamp/readme.md]: You are not explaining `block_period` here.

File: srml/timestamp/readme.md (hunk 1 above)
Comment (joepetrowski) [reply srml/timestamp/readme.md]: I think the idea of this section is so the user is at least aware of what is going in storage as a result of using this module. I guess we need to decide, is this so the user knows:

* What is going into storage, or
* What storage items are available from outside this module?

File: srml/timestamp/readme.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+Get the current time for the current block. If this function is called prior the setting the timestamp, it will return the timestamp of the previous block.
+
+Returns the timestamp as `Moment`.
+
+### Public Mutable functions (changing state)
+
+#### set(origin, now: T::Moment)
+
+Sets the current time. Extrinsic with this call should be placed at the specific position in the each block (specified by the Trait::TIMESTAMP_SET_POSITION) typically at the start of the each block. This call should be invoked exactly once per block. It will panic at the finalization phase, if this call hasn't been invoked by that time.
```
Comment (bkchr) [inline srml/timestamp/readme.md]: "specified by the Trait::TIMESTAMP_SET_POSITION" this is not correct anymore, the inherints are just at the beginning of a block.

File: srml/timestamp/readme.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+The timestamp module defines and implements the trait `TimestampInherentData` for `InherentData` to query the timestamp inherent using the timestamp `InherentIdentifier`.
+
+## Usage
+
+The following example show how to use the timestamp module in your custom module to query the current timestamp.
+
+1. Import the `timestamp` module and derive your module configuration trait with the timestamp trait.
+
+```
+use {timestamp};
```
Comment (bkchr) [inline srml/timestamp/readme.md]:
```suggestion
use timestamp;
```

File: srml/timestamp/src/lib.rs [hunk 5]
```diff
@@ -98,11 +98,16 @@ impl TimestampInherentData for InherentData {
 pub struct InherentDataProvider;
 
 #[cfg(feature = "std")]
+/// Implementation of ProvideInherentData
```
Comment (bkchr) [inline srml/timestamp/src/lib.rs]: You repeating what the code already tells you.

File: srml/timestamp/src/lib.rs [hunk 6]
```diff
@@ -98,11 +98,16 @@ impl TimestampInherentData for InherentData {
...
 impl ProvideInherentData for InherentDataProvider {
+	/// returns the identifier (storage key) for timestamp inherent data
```
Comment (bkchr) [inline srml/timestamp/src/lib.rs]:
```suggestion
	/// Returns the identifier for timestamp in inherent data
```

File: srml/timestamp/src/lib.rs [hunk 7]
```diff
@@ -98,11 +98,16 @@ impl TimestampInherentData for InherentData {
...
 	fn inherent_identifier(&self) -> &'static InherentIdentifier {
 		&INHERENT_IDENTIFIER
 	}
 
+	/// provides the current system timestamp (in seconds from epoch) as timestamp inherent data
```
Comment (bkchr) [inline srml/timestamp/src/lib.rs]: You should not repeat the implementation of the function as documentation.

And please start with capital letters at the beginning of a documentation.

File: srml/timestamp/src/lib.rs [hunk 8]
```diff
@@ -176,6 +183,7 @@ decl_module! {
 			<T::OnTimestampSet as OnTimestampSet<_>>::on_timestamp_set(now);
 		}
 
+		/// checks if the timestamp was set in the current block
```
Comment (bkchr) [inline srml/timestamp/src/lib.rs]: This comment will be shown above the implementation of a trait method in the documentation. And it repeats again the code.

File: srml/timestamp/readme.md (hunk 1 above)
Comment (gavofyork) [reply srml/timestamp/readme.md]: We don't need to (and shouldn't) be documenting stuff that is internal to this module. Only interface is important.

File: srml/timestamp/readme.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+#### set(origin, now: T::Moment)
+
+Sets the current time. Extrinsic with this call should be placed at the specific position in the each block (specified by the Trait::TIMESTAMP_SET_POSITION) typically at the start of the each block. This call should be invoked exactly once per block. It will panic at the finalization phase, if this call hasn't been invoked by that time. 
+The timestamp should be greater than the previous one by the amount specified by `block_period`.
+
+##### Errors:
+
+* Timestamp must be updated only once in the block
+* Timestamp must increment by at least <BlockPeriod> between sequential blocks
```
Comment (gavofyork) [inline srml/timestamp/readme.md]: Should be written:

```
`BlockPeriod`
```

File: srml/timestamp/readme.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+pub trait Trait: timestamp::Trait { }
+```
+
+2. In your module call the timestamp module's get function to get the current timestamp,
+
+```
+let now = <timestamp::Module<T>>::get();
+```
+
+## Implementation Details
```
Comment (gavofyork) [inline srml/timestamp/readme.md]: Avoid documenting implementation - this is user-documentation and should be restricted to interface.

File: srml/timestamp/src/lib.rs [hunk 11]
```diff
@@ -211,24 +221,37 @@ impl<T: Trait> Module<T> {
...
+	/// Sets the new timestamp by calling the `set` dispatchable function
 	fn create_inherent(data: &InherentData) -> Option<Self::Call> {
 		let data = extract_inherent_data(data).expect("Gets and decodes timestamp inherent data");
 
 		let next_time = cmp::max(As::sa(data), Self::now() + Self::block_period());
 		Some(Call::set(next_time.into()))
 	}
 
+	/// Allows other validators to validate the inherent data for timestamp
+	/// Checks if the timestamp is not too far in the future (current + max drift)
```
Comment (gavofyork) [inline srml/timestamp/src/lib.rs]: Use backticks to escape formulae

File: srml/timestamp/src/lib.rs [hunk 12]
```diff
@@ -211,24 +221,37 @@ impl<T: Trait> Module<T> {
...
+/// gets the inherent against the timestamp INHERENT_IDENTIFIER
+/// tries to deserialize the timestamp inherent into the InherentType (u64)
 fn extract_inherent_data(data: &InherentData) -> Result<InherentType, RuntimeString> {
 	data.get_data::<InherentType>(&INHERENT_IDENTIFIER)
 		.map_err(|_| RuntimeString::from("Invalid timestamp inherent data encoding."))?
 		.ok_or_else(|| "Timestamp inherent data is not provided.".into())
 }
 
+/// Implementation of `ProvideInherent` trait for timestamp module
+/// create_inherent - to set a new timestamp by the block proposer
```
Comment (gavofyork) [inline srml/timestamp/src/lib.rs]: Use backticks for identifiers.

File: srml/timestamp/src/lib.rs [hunk 13]
```diff
@@ -211,24 +221,37 @@ impl<T: Trait> Module<T> {
 	}
 }
 
+/// extracts the timestamp inherent from inherent data
```
Comment (gavofyork) [inline srml/timestamp/src/lib.rs]: Capitalise first word

File: srml/timestamp/src/lib.rs [hunk 14]
```diff
@@ -211,24 +221,37 @@ impl<T: Trait> Module<T> {
...
+/// gets the inherent against the timestamp INHERENT_IDENTIFIER
```
Comment (gavofyork) [inline srml/timestamp/src/lib.rs]: Backticks for identifiers (`INHERENT_IDENTIFIER`, `InherentType`, `u64`)

File: srml/timestamp/readme.md [hunk 15]
```diff
@@ -0,0 +1,81 @@
+# Timestamp Module
```
Comment (ltfschoen) [inline srml/timestamp/readme.md]: i think for consistency with the rest of the codebase that the actual readme file should be named to README.adoc (AsciiDoc format) instead of being in Markdown format. The name of the file should be `README` not lowercase `readme`

File: srml/timestamp/readme.md [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+Returns the timestamp as `Moment`.
+
+### Public Mutable functions (changing state)
+
+#### set(origin, now: T::Moment)
+
+Sets the current time. Extrinsic with this call should be placed at the specific position in the each block (specified by the Trait::TIMESTAMP_SET_POSITION) typically at the start of the each block. This call should be invoked exactly once per block. It will panic at the finalization phase, if this call hasn't been invoked by that time. 
+The timestamp should be greater than the previous one by the amount specified by `block_period`.
+
+##### Errors:
```
Comment (ltfschoen) [inline srml/timestamp/readme.md]: should this instead be `### Errors:` (i.e. a subheading of "Public Interface" instead of being a subheading of the "set" method)

File: srml/timestamp/readme.md [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+
+* `Moment` - Represents the current timestamp. Generally implemented as a `u64`.
+
+### Storage Items
+
+* `Now`: `Moment` - The current timestamp. A `u64` having the timestamps as **total seconds from the unix epoch**.
+* `BlockPeriod`: `Moment` - The minimum (and advised) period between blocks. Also represented as `Moment`.
+* `DidUpdate`: `bool` - Private flag to check if the timestamp was updated for the current block.
+
+### Public Inspection functions - Immutable (getters)
```
Comment (ltfschoen) [inline srml/timestamp/readme.md]: could we make these headings more concise by renaming them to just "Public Getters" and "Public Setters"

File: srml/timestamp/readme.md [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+
+The timestamp module defines and implements the trait `TimestampInherentData` for `InherentData` to query the timestamp inherent using the timestamp `InherentIdentifier`.
+
+## Usage
+
+The following example show how to use the timestamp module in your custom module to query the current timestamp.
+
+1. Import the `timestamp` module and derive your module configuration trait with the timestamp trait.
+
+```
```
Comment (ltfschoen) [inline srml/timestamp/readme.md]: suggest adding ````rust`

File: srml/timestamp/readme.md [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+
+* `Now`: `Moment` - The current timestamp. A `u64` having the timestamps as **total seconds from the unix epoch**.
+* `BlockPeriod`: `Moment` - The minimum (and advised) period between blocks. Also represented as `Moment`.
+* `DidUpdate`: `bool` - Private flag to check if the timestamp was updated for the current block.
+
+### Public Inspection functions - Immutable (getters)
+
+#### get()
+
+Get the current time for the current block. If this function is called prior the setting the timestamp, it will return the timestamp of the previous block.
```
[the remark this answers is not in the document]
Comment (ltfschoen) [reply srml/timestamp/readme.md]: this hasn't been addressed yet

File: srml/timestamp/readme.md [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+The timestamp should be greater than the previous one by the amount specified by `block_period`.
+
+##### Errors:
+
+* Timestamp must be updated only once in the block
+* Timestamp must increment by at least <BlockPeriod> between sequential blocks
+
+### Inherent Data
+
+The timestamp module manages the block timestamp using InherentData. To identify the timestamp inherent, it defines an `InherentIdentifier`.
```
[the remark this answers is not in the document]
Comment (ltfschoen) [reply srml/timestamp/readme.md]: this hasn't been resolved yet

File: srml/timestamp/README.md [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+
+The timestamp module manages the block timestamp using InherentData. To identify the timestamp inherent, it defines an `InherentIdentifier`.
+
+The timestamp module defines and implements the trait `TimestampInherentData` for `InherentData` to query the timestamp inherent using the timestamp `InherentIdentifier`.
+
+## Usage
+
+The following example shows how to use the timestamp module in your custom module to query the current timestamp.
+
+In your custom module, after importing the `timestamp` module and deriving your module's configuration trait with the timestamp trait, call the timestamp module's `get` function to get the current timestamp,
```
Comment (ltfschoen) [inline srml/timestamp/README.md]: paragraph should end with full stop instead of comma

File: srml/timestamp/readme.md (hunk 16 above)
Comment (ltfschoen) [reply srml/timestamp/readme.md]: haven't seen any change or response to this

File: srml/timestamp/readme.md (hunk 17 above)
Comment (ltfschoen) [reply srml/timestamp/readme.md]: it looks like they've changed to `Public Immutable functions` and `Public Mutable functions`, which differs from the example template https://github.com/paritytech/substrate/pull/1947/files

File: srml/timestamp/readme.md (hunk 18 above)
Comment (ltfschoen) [reply srml/timestamp/readme.md]: i just did a search for triple backslash and found one missing `rust`

File: srml/timestamp/readme.md (hunk 16 above)
Comment (gautamdhameja) [reply srml/timestamp/readme.md]: This was left as-is because the template has errors for each function.

File: srml/timestamp/readme.md (hunk 17 above)
Comment (gautamdhameja) [reply srml/timestamp/readme.md]: I found it better this way, will update the template instead.

File: srml/timestamp/readme.md (hunk 18 above)
Comment (gautamdhameja) [reply srml/timestamp/readme.md]: Not sure what is needed here. This text is removed though, to address another review comment.

File: srml/timestamp/src/lib.rs [hunk 22]
```diff
@@ -116,6 +171,7 @@ impl ProvideInherentData for InherentDataProvider {
 			})
 	}
 
+	/// Tries to convert the `InherentError` into a string
```
Comment (thiolliere) [inline srml/timestamp/src/lib.rs]: Indeed I'm in favor of documentating all public stuff even obvious ones, but on the other end I'm not necessarily in favor of documentating private function when name is expressful enough

**Comment by bkchr** [comment]
> in rustdoc to make link you can write [`NameOfTheStuff`] and then at the end of the doc after an empty line you write [`NameOfTheStuff`]: struct.NameOfTheStuff.html

You don't need to do this anymore, just write [`NameOfTheStuff`] and it works. Rustdocs supports intra doc links for some time now.

File: srml/timestamp/src/lib.rs [hunk 23]
```diff
@@ -14,21 +14,74 @@
...
+//! 
+//! The following example shows how to use the timestamp module in your custom module to query the current timestamp.
+//! 
+//! ### Prerequisites
+//! 
+//! Import the `timestamp` module in your custom module.
+//! 
+//! ```ignore
+//! use timestamp;
+//! ```
```
[the remark this answers is not in the document]
Comment (bkchr) [reply srml/timestamp/src/lib.rs]: Yeah, not necessary with rust 2018 and the full name would anyway be `srml-timestamp`.

File: srml/timestamp/src/lib.rs [hunk 24]
```diff
@@ -14,21 +14,74 @@
...
+//! In your custom module, after importing the `timestamp` module and deriving your module's configuration trait with the timestamp trait, 
+//! call the timestamp module's `get` function to get the current timestamp.
+//! 
+//! ```ignore
+//! let now = <timestamp::Module<T>>::get();
+//! ```
+//! 
+//! ### Example from SRML
+//! 
+//! The [`Session` module](https://github.com/paritytech/substrate/blob/master/srml/session/src/lib.rs) uses the `timestamp` module for session management.
```
[the remark this answers is not in the document]
Comment (gautamdhameja) [reply srml/timestamp/src/lib.rs]: I am avoiding linking to line numbers because they can change more frequently (updates, bug fixes, new code, etc.)

**Comment by thiolliere** [comment]
> > in rustdoc to make link you can write [`NameOfTheStuff`] and then at the end of the doc after an empty line you write [`NameOfTheStuff`]: struct.NameOfTheStuff.html
> 
> You don't need to do this anymore, just write [`NameOfTheStuff`] and it works. Rustdocs supports intra doc links for some time now.

it seems to be the case on nightly only

File: srml/timestamp/src/lib.rs [hunk 25]
```diff
@@ -14,21 +14,59 @@
...
+//! **NOTE:** The timestamp module is the recommended way to query the on-chain time instead of using an approach based on block numbers. 
+//! The block numbers based time measurement can cause issues because of cummulative calculation errors and hence it should be avoided.
+//! 
+//! ## Interface
+//! 
+//! ### Dispatchable functions
+//! 
+//! * `set` - Sets the current time.
+//! 
+//! Please refer to the [`Call`] enum and its associated functions for a detailed list of dispatchable functions.
```
Comment (gavofyork) [inline srml/timestamp/src/lib.rs]: Why is this here? It's always true.

File: srml/timestamp/src/lib.rs [hunk 26]
```diff
@@ -211,6 +251,7 @@ impl<T: Trait> Module<T> {
 	}
 }
 
+/// Extracts the timestamp inherent from inherent data.
```
Comment (gavofyork) [inline srml/timestamp/src/lib.rs]: says nothing that the code doesn't.

File: srml/timestamp/src/lib.rs [hunk 27]
```diff
@@ -14,21 +14,57 @@
...
+//! 
+//! ### Dispatchable functions ([`Call`])
+//! 
+//! * `set` - Sets the current time.
+//! 
+//! ### Public functions
+//! 
+//! * `get` - Gets the current time for the current block. If this function is called prior the setting to timestamp, it will return the timestamp of the previous block.
+//! 
+//! * `block_period` - Gets the block period for the chain.
```
Comment (shawntabrizi) [inline srml/timestamp/src/lib.rs]: Nitpick: Could we be just a little more clear here?

"Gets the minimum (and advised) period between blocks for the chain"

File: srml/timestamp/src/lib.rs [hunk 28]
```diff
@@ -14,21 +14,74 @@
...
+//! ```
+//! 
+//! ### Simple getter
+//! 
+//! In your custom module, after importing the `timestamp` module and deriving your module's configuration trait with the timestamp trait, 
+//! call the timestamp module's `get` function to get the current timestamp.
+//! 
+//! ```ignore
+//! let now = <timestamp::Module<T>>::get();
+//! ```
```
[the remark this answers is not in the document]
Comment (shawntabrizi) [reply srml/timestamp/src/lib.rs]: I agree, I can try to write an MVCE and include it here for you to update with

[the remark this answers is not in the document]
Comment (gautamdhameja) [reply srml/timestamp/src/lib.rs]: I had a more detailed example before but @bkchr suggested to remove the obvious parts and reduce the text. See here 
 - https://github.com/paritytech/substrate/blob/ac865a8bbfeeab831798ced3b6ba9e9ffad74dba/srml/timestamp/src/lib.rs#L49

File: srml/timestamp/src/lib.rs [hunk 29]
```diff
@@ -14,21 +14,57 @@
...
+//! 
+//! Import the `timestamp` module in your custom module and derive the module configuration trait from the `timestamp` trait.
+//! 
+//! ### Get current timestamp
+//! 
+//! ```ignore
+//! let now = <timestamp::Module<T>>::get();
+//! ```
+//! 
+//! ### Example usage
```
Comment (ltfschoen) [inline srml/timestamp/src/lib.rs]: Should the title be "Example from SRML" as shown here https://github.com/paritytech/substrate-developer-hub/issues/44

File: srml/timestamp/src/lib.rs [hunk 30]
```diff
@@ -14,21 +14,57 @@
...
+//! 
+//! The [`Session` module](https://github.com/paritytech/substrate/blob/master/srml/session/src/lib.rs) uses the `timestamp` module for session management.
+//!
```
Comment (ltfschoen) [inline srml/timestamp/src/lib.rs]: Do we need a Genesis Config / Dependencies / Reference sections as mentioned here https://github.com/paritytech/substrate-developer-hub/issues/44

File: srml/timestamp/src/lib.rs (hunk 28 above)
[the remark this answers is not in the document]
Comment (shawntabrizi) [reply srml/timestamp/src/lib.rs]: Okay, that seems agreeable too, but it seems like for usage you should at least include explicitly that you need to inherent from the `timestamp::Trait` and how to inherent from it.

File: srml/timestamp/src/lib.rs (hunk 30 above)
Comment (gautamdhameja) [reply srml/timestamp/src/lib.rs]: I think it does not need these sections. We do not have anything specific to tell there.

File: srml/timestamp/src/lib.rs (hunk 28 above)
[the remark this answers is not in the document]
Comment (gautamdhameja) [reply srml/timestamp/src/lib.rs]: I think the above MVCE is good and concise. I have added it back. It is also good for consistency.

**Comment by shawntabrizi** [comment]
🥇
